### ISSUE TITLE
Type Variable ranging over a union behaves inconsistently between int and float and with mypy

### ISSUE THREAD

**Issue by maegul** [microsoft/pyright#3133]
**Describe the bug**
`Int` and `float` variables show inconsistent behaviour when typed relative to a type variable that ranges over a Union of `int` and `float` (`int | float`) and another type (`numpy.ndarray`).  Specifically, `int` variables are invalid, `float` are valid, while `mypy` reports no issues in either case.  Issue may be between keyboard and chair on this one.

**To Reproduce**
```python
from typing import TypeVar, Generic, Union
from dataclasses import dataclass
import numpy as np

scalar = Union[int, float]
scal_arr = TypeVar('scal_arr', scalar, np.ndarray)

@dataclass
class Data(Generic[scal_arr]):
    data: scal_arr

# return only ints | arrays
def test_generic_data_int(a: Data[scal_arr]) -> scal_arr:
    if isinstance(a.data, (int, float)):
        value = int(a.data / 3)  # not OK in pyright (fine in mypy)
    else:
        value = a.data.astype(int)

    # pyright:  int | ndarray*
    # mypy:     int* | ndarray*
    reveal_type(value)

    # pyright:  int | ndarray* cannot be assiged to return type scal_arr@test_generic_data
    # mypy:     :)
    return value  # not OK (for pyright)

# return only floats | arrays
def test_generic_data_float(a: Data[scal_arr]) -> scal_arr:
    if isinstance(a.data, (int, float)):
        value = a.data / 3
    else:
        value = a.data.astype(float)

    # pyright:  float* | ndarray*
    # mypy:     float | ndarray*
    reveal_type(value)

    # pyright:  :)
    # mypy:     :)
    return value  # OK

```

**Expected behavior**

* Consistency between `mypy` and `pyright`.
* Consistency between `int` and `float`.
    - It's not clear to me (perhaps/likely because of ignorance) why `float` variables would be valid and `int` not.
    - _My guess would_ be that `scalar` is somewhere being reverted to the least specific of its components (see `PEP 483`), ie `float`, due to the pragmatic `de facto` type hierarchy of `complex > float > int` and some type variable `invariance` logic.  The inconsistency with `mypy` indicates to me that this is an edge case of what exactly this pragmatic typing policy entails and that either checker might be at fault ... ???
* _Personally_, by my own (incomplete) understanding, both `int` and `float` ought to satisfy `scal_arr` when `a.data` is either an `int` or `float`.
* _But_, given the type-checking of the code snippet below that is similar in structure but with different types, I'd expect neither `int` or `float` to be valid whenever the variable's type constantly reduced to either type.
    - Of course I can only point back to the pragmatics around `int` and `float` to explain why `mypy` is unhappy in this case, though `mypy`'s issue seems to be more around parsing the conditional code, which is not an issue I am not well versed in with respect to type checkers.

```python
class A: pass
class B: pass

basic_T = Union[A, B]
basic_arr = TypeVar('basic_arr', basic_T, np.ndarray)

@dataclass
class basicData(Generic[basic_arr]):
    data: basic_arr

def test_basic_type(x: basicData[basic_arr]) -> basic_arr:

    if isinstance(x.data, (A, B)):
        value = B()
    elif isinstance(x.data, np.ndarray):
        value = np.ones(10)
    else:  # necessary for some reason ??
        value = x.data

    # pyright:  B | ndarray cannot be assiged to return type basic_arr@test_basic_type
    # mypy:     error: Incompatible return value type (got "B", expected "ndarray")  ???
    return value
```

**VS Code extension or command-line**
* `pyright`: `1.1.143` (through `Sublime Text` plugin/extension `sublimelsp:LSP-pyright`)
* `mypy`: `0.931`

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.227, which I just published. It will also be included in the next release of pylance.
